File: src/main/java/de/com/fdm/twitch/tmi/TmiService.java
package de.com.fdm.twitch.tmi;

import com.github.philippheuer.credentialmanager.domain.OAuth2Credential;
import com.github.twitch4j.TwitchClient;
import com.github.twitch4j.TwitchClientBuilder;
import com.github.twitch4j.chat.events.channel.ChannelMessageActionEvent;
import com.github.twitch4j.chat.events.channel.ChannelMessageEvent;
import de.com.fdm.bot.MessageHandler;
import de.com.fdm.lib.Result;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

@Service
public class TmiService {
    private final TwitchClient client;
    private final MessageHandler messageHandler;

    public TmiService(
            @Value("${bot.auth}") String botAuth,
            @Autowired MessageHandler messageHandler
    ) {
        this.messageHandler = messageHandler;

        OAuth2Credential credentials = new OAuth2Credential("twitch", botAuth);
        this.client = TwitchClientBuilder.builder()
                .withEnableChat(true)
                .withChatAccount(credentials)
                .build();

        client.getEventManager().onEvent(ChannelMessageEvent.class, this::handleMessage);
        client.getEventManager().onEvent(ChannelMessageActionEvent.class, this::handleActionMessage);
    }

    private void handleActionMessage(ChannelMessageActionEvent event) {
        ChannelMessageEvent msgEvent = new ChannelMessageEvent(
                event.getChannel(),
                event.getMessageEvent(),
                event.getUser(),
                event.getMessage(),
                event.getPermissions()
        );

        handleMessage(msgEvent);
    }


    private void handleMessage(ChannelMessageEvent event) {
        InboundMessage msg = new InboundMessage(event);

        Result<OutboundMessage, String> result = messageHandler.handleMessage(msg);
        if (result.isOk()) {
            send(result.getValue());
        }
    }

    public void send(OutboundMessage msg) {
        client.getChat().sendMessage(msg.getChannel(), msg.getText());
    }

    public void joinChannels(String[] channels) {
        for (String channel : channels) {
            client.getChat().joinChannel(channel);
        }
    }
}

File: src/main/java/de/com/fdm/twitch/tmi/TmiService.kt
package de.com.fdm.twitch.tmi

import com.github.philippheuer.credentialmanager.domain.OAuth2Credential
import com.github.twitch4j.TwitchClient
import com.github.twitch4j.TwitchClientBuilder
import com.github.twitch4j.chat.events.channel.ChannelMessageEvent
import de.com.fdm.bot.api.haste.HasteService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import java.util.function.Consumer

@Service
class TmiService(
        @Value("\${bot.auth}") botAuth: String,
        @param:Autowired private val hasteService: HasteService
) {
    private val client: TwitchClient

    init {
        val credentials = OAuth2Credential("twitch", botAuth)
        client = TwitchClientBuilder.builder()
                .withEnableChat(true)
                .withChatAccount(credentials)
                .build()
    }

    fun setCallback(callback: Consumer<TmiMessage>) {
        client.eventManager.onEvent(ChannelMessageEvent::class.java) {
            event: ChannelMessageEvent -> callback.accept(TmiMessage(event))
        }
    }

    fun send(channel: String?, msg: String) {
        if (msg.length >= 500) {
            val url = hasteService.upload(msg)
            client.chat.sendMessage(channel, url)
            return
        }

        client.chat.sendMessage(channel, msg)
    }

    fun joinChannels(channels: Array<String?>) {
        for (channel in channels) {
            client.chat.joinChannel(channel)
        }
    }

    fun joinChannel(channel: String?) {
        client.chat.joinChannel(channel)
    }

    fun partChannel(channel: String?) {
        client.chat.leaveChannel(channel)
    }

    fun getChannels() : Set<String> {
        return client.chat.channels
    }
}